### PR TITLE
[Impeller] eliminate sub-render pass for blended color + texture vertices.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/145707

Part of https://github.com/flutter/flutter/issues/131345

If we're drawing vertices with per-color, a non-advanced blend, and a texture (with or without coordinates), use the porter duff shader to perform the blend without creating a sub render pass. IN addition to being more performant, this eliminates any potential rendering bugs caused by overlapping vertices.

This is not yet fixed for advanced blends.

### CODE REVIEW

File: impeller/aiks/canvas.cc [hunk 1]
```diff
@@ -885,6 +889,28 @@ void Canvas::DrawVertices(const std::shared_ptr<VerticesGeometry>& vertices,
     return;
   }
 
+  // If there is are per-vertex colors, an image, and the blend mode
+  // is simple we can draw without a sub-renderpass.
+  if (blend_mode <= BlendMode::kModulate && vertices->HasVertexColors() &&
```
Comment (jonahwilliams) [inline impeller/aiks/canvas.cc:894]: We can do this for the rest of the blend modes once I make some updates to the advanced blends. After that, the only time we'd use the old path is for a gradient/shader color source. There is probably an easier refactor there to eagerly convert those to a texture and then let them go through this code path instead.

File: impeller/entity/geometry/vertices_geometry.cc [hunk 2]
```diff
@@ -243,6 +244,64 @@ GeometryResult VerticesGeometry::GetPositionUVBuffer(
...
+  using VS = PorterDuffBlendPipeline::VertexShader;
+
+  auto index_count = indices_.size();
+  auto vertex_count = vertices_.size();
+  auto uv_transform =
+      texture_coverage.GetNormalizingTransform() * effect_transform;
+  auto has_texture_coordinates = HasTextureCoordinates();
+
+  size_t total_vtx_bytes = vertices_.size() * sizeof(VS::PerVertexData);
+  size_t total_idx_bytes = index_count * sizeof(uint16_t);
```
Comment (gaaclarke) [inline impeller/entity/geometry/vertices_geometry.cc:262]: Can you move this down to where it's used please?

File: impeller/entity/shaders/blending/porter_duff_blend.frag [hunk 3]
```diff
@@ -36,13 +36,22 @@ f16vec4 Sample(f16sampler2D texture_sampler, vec2 texture_coords) {
   return IPHalfSampleDecal(texture_sampler, texture_coords);
 }
 
+f16vec4 IPHalfClamp(f16vec4 color) {
```
Comment (gaaclarke) [inline impeller/entity/shaders/blending/porter_duff_blend.frag:39]:
```suggestion
f16vec4 IPHalfClamp01(f16vec4 color) {
```

File: impeller/entity/shaders/blending/porter_duff_blend.frag [hunk 4]
```diff
@@ -36,13 +36,22 @@ f16vec4 Sample(f16sampler2D texture_sampler, vec2 texture_coords) {
...
+
 void main() {
   f16vec4 dst =
       texture(texture_sampler_dst, v_texture_coords) * frag_info.input_alpha;
   f16vec4 src = v_color;
-  frag_color =
+  // This currently needs a clamp so that floating point textures blend
+  // correctly in wide gamut. Remove if we switch to a fixed point extended
+  // range format.
+  frag_color = IPHalfClamp(
```
Comment (gaaclarke) [inline impeller/entity/shaders/blending/porter_duff_blend.frag:52]: I not certain we have to do this.  I'd rather be more pragmatic with it since it is very complicated.  We have wide gamut tests now so we can demonstrate when/where this goes wrong.  If we get a case that is broken we could use push constants or uniforms to specify the clamping range.  We'd need that for the fixed point format as well.

At a minimum it should be in its own PR.

File: impeller/aiks/aiks_unittests.cc [hunk 5]
```diff
@@ -3227,6 +3227,39 @@ TEST_P(AiksTest, VerticesGeometryUVPositionDataWithTranslate) {
...
+  std::vector<uint16_t> indices = {0u, 1u, 2u, 3u, 4u, 5u};
+  std::vector<Point> texture_coordinates = {};
+  std::vector<Color> vertex_colors = {
+      Color::Red().WithAlpha(0.5),   Color::Blue().WithAlpha(0.5),
+      Color::Green().WithAlpha(0.5), Color::Red().WithAlpha(0.5),
+      Color::Blue().WithAlpha(0.5),  Color::Green().WithAlpha(0.5),
+  };
+  auto geometry = std::make_shared<VerticesGeometry>(
+      vertices, indices, texture_coordinates, vertex_colors,
+      Rect::MakeLTRB(0, 0, 1, 1), VerticesGeometry::VertexMode::kTriangleStrip);
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/aiks/aiks_unittests.cc]: Good catch, I forgot to supply the primitive type to the pipeline.

File: impeller/aiks/canvas.cc [hunk 6]
```diff
@@ -885,6 +889,28 @@ void Canvas::DrawVertices(const std::shared_ptr<VerticesGeometry>& vertices,
...
+      paint.color_source.GetType() == ColorSource::Type::kImage) {
+    auto contents = std::make_shared<VerticesSimpleBlendContents>();
+    contents->SetBlendMode(blend_mode);
+    contents->SetAlpha(paint.color.alpha);
+    contents->SetGeometry(vertices);
+
+    auto color_contents = std::reinterpret_pointer_cast<TiledTextureContents>(
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/aiks/canvas.cc]: Switched to static cast.

The cast is safe as long as it is guared by the paint.color_source.GetType() == ColorSource::Type::kImage above. I would actually like to combine the color sources into a std::variant so we can lose the heap allocation, that would let us have a visitor

File: impeller/entity/contents/vertices_contents.cc [hunk 7]
```diff
@@ -198,4 +225,120 @@ bool VerticesColorContents::Render(const ContentContext& renderer,
...
+  dst_sampler_descriptor.width_address_mode =
+      TileModeToAddressMode(tile_mode_x_, renderer.GetDeviceCapabilities())
+          .value_or(SamplerAddressMode::kDecal);
+  dst_sampler_descriptor.height_address_mode =
+      TileModeToAddressMode(tile_mode_y_, renderer.GetDeviceCapabilities())
+          .value_or(SamplerAddressMode::kDecal);
+
+  if (renderer.GetDeviceCapabilities().SupportsDecalSamplerAddressMode()) {
+    dst_sampler_descriptor.width_address_mode = SamplerAddressMode::kDecal;
+    dst_sampler_descriptor.height_address_mode = SamplerAddressMode::kDecal;
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/entity/contents/vertices_contents.cc]: Done

File: impeller/entity/geometry/vertices_geometry.cc (hunk 2 above)
Comment (jonahwilliams) [reply impeller/entity/geometry/vertices_geometry.cc]: Done

File: impeller/entity/shaders/blending/porter_duff_blend.frag (hunk 3 above)
Comment (jonahwilliams) [reply impeller/entity/shaders/blending/porter_duff_blend.frag]: Removed

File: impeller/entity/shaders/blending/porter_duff_blend.frag (hunk 4 above)
Comment (jonahwilliams) [reply impeller/entity/shaders/blending/porter_duff_blend.frag]: Removed

Comment (jonahwilliams) [reply impeller/entity/shaders/blending/porter_duff_blend.frag]: If we don't clamp the alpha, then this change re-breaks the vertices, so we need it otherwise this change is unsafe.

File: impeller/aiks/canvas.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/aiks/canvas.cc:894]: The fast path is:

1. If its a solid color and nothing else, just use a solid color contents.
2. If its just a texture, or a texture and texture coordinates, just use tiled texture contents.
3. This path.
4. everything else, hell on earth

File: impeller/entity/shaders/blending/porter_duff_blend.frag (hunk 4 above)
Comment (gaaclarke) [reply impeller/entity/shaders/blending/porter_duff_blend.frag]: Yea, clamping the alpha sounds good.  The jury is still out wether clamping the color components is the correct behavior, I'm looking into that.  Can you add a TODO linking that BGRA10_XR issue above the clamp.

I'm not sure where this shader is used fully, but clamping the color components may well break our wide gamut devicelab tests.

Comment (jonahwilliams) [reply impeller/entity/shaders/blending/porter_duff_blend.frag]: I'm just going to clamp the alpha, draw atlas is rendering wrong still at ToT.

https://github.com/flutter/engine/pull/51792

File: impeller/aiks/canvas.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (gaaclarke) [reply impeller/aiks/canvas.cc]: Can we add a `bool Contents::IsTiledTextureContents` with a TODO for the std::variant?

My concern is that while `paint.color_source.GetType() == ColorSource::Type::kImage` may be a proper guard now, someone reading the code cannot know that it is a proper guard and we have no backstop for future changes that may invalidate that guarantee.

A bad cast in C++ is undefined behavior, it won't fail gracefully like in other languages.

[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/aiks/canvas.cc]: If we guarding against arbitrary future code changes, then i could also change IsTiledTextureContents() to return true for a different type.

**Comment by jonahwilliams** [comment]
Golden currently looks wrong for metal, investigating.

File: impeller/aiks/canvas.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (gaaclarke) [reply impeller/aiks/canvas.cc]: > If we guarding against arbitrary future code changes

I wouldn't say "guarding".  What we are building with our tests is the confidence to change the code and know that it is still correct after a change.  A downcast that is determined to be safe from a non-local derivative isn't helping us.

We agree the visitor or std::variant would be better solutions.  That's why I think we should have the TODO.  If you want to make that change instead that'd be fine with me.  I assumed that was out of scope, `IsTiledTextureContents` seems the easiest solution.  It's ugly, but downcasting is ugly and we have a way to make it better if we want.  I don't think there is another alternative to assert this is a safe downcast, we could enable RTTI for debug builds of the test runner, but I think it's best if we keep the test runners executing like how they will in production.

[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/aiks/canvas.cc]: Its not that its ugly @gaaclarke , its just that I don't believe it provides any additional verification.

[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/aiks/canvas.cc]: I'll do this the right way after https://github.com/flutter/engine/pull/51853 , but I still need to figure out why the golden looks wrong on metal.

**Comment by jonahwilliams** [comment]
Done, golden problem was mismatched colors/vertices (the assers are in dart:ui)

File: impeller/aiks/canvas.cc [hunk 8]
```diff
@@ -885,6 +929,26 @@ void Canvas::DrawVertices(const std::shared_ptr<VerticesGeometry>& vertices,
     return;
   }
 
+  // If there is are per-vertex colors, an image, and the blend mode
+  // is simple we can draw without a sub-renderpass.
+  if (blend_mode <= BlendMode::kModulate && vertices->HasVertexColors()) {
+    if (auto maybe_image_data = GetImageColorSourceData(paint.color_source)) {
+      auto image_data = maybe_image_data.value();
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply impeller/aiks/canvas.cc:936]: Yes

**Comment by jonahwilliams** [comment]
Nice performance bump from this: https://flutter-flutter-perf.skia.org/e/?queries=device_type%3DPixel_7_Pro%26sub_result%3D90th_percentile_frame_rasterizer_time_millis%26sub_result%3D99th_percentile_frame_rasterizer_time_millis%26sub_result%3Daverage_frame_rasterizer_time_millis%26sub_result%3Dworst_frame_rasterizer_time_millis%26test%3Ddraw_vertices_perf__timeline_summary&selected=commit%3D40161%26name%3D%252Carch%253Dintel%252Cbranch%253Dmaster%252Cconfig%253Ddefault%252Cdevice_type%253DPixel_7_Pro%252Cdevice_version%253Dnone%252Chost_type%253Dlinux%252Csub_result%253D99th_percentile_frame_rasterizer_time_millis%252Ctest%253Ddraw_vertices_perf__timeline_summary%252C

**Comment by gaaclarke** [comment]
> Nice performance bump from this

The graph goes down, not up.  I'd call it more of a dump than a bump.
